File: CLDemo/Others/CLLanguageManager/CLLanguageManager.m
//
//  CLLanguageManager.m
//  CLDemo
//
//  Created by AUG on 2018/11/7.
//  Copyright © 2018年 JmoVxia. All rights reserved.
//

#import "CLLanguageManager.h"

static NSString *CLUserLanguageKey = @"CLUserLanguageKey";


@implementation CLLanguageManager

+ (void)setUserLanguage:(NSString *)userLanguage {
    //跟随手机系统
    if (!userLanguage.length) {
        [self resetSystemLanguage];
        return;
    }
    //用户自定义
    [[NSUserDefaults standardUserDefaults] setValue:userLanguage forKey:CLUserLanguageKey];
    [[NSUserDefaults standardUserDefaults] setValue:@[userLanguage] forKey:@"AppleLanguages"];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

+ (NSString *)userLanguage {
    return [[NSUserDefaults standardUserDefaults] valueForKey:CLUserLanguageKey];
}

/**
 重置系统语言
 */
+ (void)resetSystemLanguage {
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:CLUserLanguageKey];
    [[NSUserDefaults standardUserDefaults] setValue:nil forKey:@"AppleLanguages"];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

@end

File: CLDemo-Swift/CLDemo-Swift/Class/CLChangeLanguageController/CLBundleManager/CLLanguageManager.swift
//
//  CLBundle.swift
//  CLDemo-Swift
//
//  Created by Chen JmoVxia on 2021/6/22.
//

import UIKit

extension CLLanguageManager {
    enum Language: String {
        ///跟随系统
        case system = ""
        ///英文
        case english = "en"
        ///简体中文
        case chineseSimplified = "zh-Hans"
    }
}

class CLLanguageManager: NSObject {
    static let shared: CLLanguageManager = CLLanguageManager()
    private (set) var bundle: Bundle = Bundle.main
    private (set) var currentLanguage: Language = .system {
        didSet {
            if currentLanguage != oldValue {
                if currentLanguage.rawValue.isEmpty {
                    self.bundle = .main
                    return
                }
                guard let bundlePath = Bundle.main.path(forResource: currentLanguage.rawValue, ofType: "lproj"),
                      let bundle = Bundle(path: bundlePath)
                else {
                    return
                }
                self.bundle = bundle
            }
        }
    }
    private override init() {
        super.init()
        defer {
            if let string = UserDefaults.standard.string(forKey: "UserLanguage"),
               let language = Language(rawValue: string) {
                self.currentLanguage = language
            }
        }
    }
}
extension CLLanguageManager {
    ///设置用户自定义语言
    static func setUserLanguage(_ language: Language) {
        UserDefaults.standard.setValue(language.rawValue, forKey: "UserLanguage")
        UserDefaults.standard.synchronize()
        
        shared.currentLanguage = language
    }
}
